pkg/apis/application/v1alpha1/app_project_types.go

```diff
@@ -356,7 +356,17 @@ func (proj AppProject) IsSourcePermitted(src ApplicationSource) bool {
 }
 
 // IsDestinationPermitted validates if the provided application's destination is one of the allowed destinations for the project
-func (proj AppProject) IsDestinationPermitted(dst ApplicationDestination) bool {
+func (proj AppProject) IsDestinationPermitted(dst ApplicationDestination, applicationName string, appsThatAllowedRunInCluster []string) bool {
+	// in case if application destination is in-cluster and allowed apps defined in argo-cm
+	if dst.Server == KubernetesInternalAPIServerAddr && len(appsThatAllowedRunInCluster) > 0 {
+		for _, appNameFromSettings := range appsThatAllowedRunInCluster {
+			if appNameFromSettings == applicationName {
+				return true
```

Comment:
if it's true here you shouldn't return yet because you skip the rest of the checks